sbin/zfsbootcfg/zfsbootcfg.8
.\" Copyright (c) 2016 Andriy Gapon
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd July 22, 2020
.Dt ZFSBOOTCFG 8
.Os
.Sh NAME
.Nm zfsbootcfg
.Nd "specify zfsboot options for the next boot"
.Sh SYNOPSIS
.Nm
.Ao Ar options Ac
.Nm
.Op Fl k Ar key
.Op Fl p
.Op Fl t Ar type
.Op Fl v Ar value
.Op Fl z Ar pool
.Nm
.Sh DESCRIPTION
.Nm
is used to set
.Xr boot.config 5 Ns -style
options to be used by
.Xr zfsboot 8 ,
.Xr gptzfsboot 8
or
.Xr loader 8
the next time the machine is booted.
Once
.Xr zfsboot 8
or
.Xr gptzfsboot 8
or
.Xr loader 8
reads the information, it is deleted.
If booting fails, the machine automatically reverts to the previous
boot configuration.
The information is stored in a special boot environment area of a ZFS pool.
.Pp
If used without arguments,
.Nm
will output the current boot configuration, if set.
.Pp
The following options are supported by
.Nm :
.Bl -tag -width indent
.It Fl k Ar key
Define key for
.Ao key , value Ac
pair.
.It Fl p
Print all information stored in ZFS pool bootenv area.
.It Fl t Ar type
Set type of
.Ar value
used in
.Ao key , value Ac
pair.
Currently, the only supported type is
.Ar DATA_TYPE_STRING .
.It Fl v Ar value
Define value for
.Ao key , value Ac
pair.
.It Fl z Ar pool
Operate on
.Ar pool .
.El
.Sh ENVIRONMENT
.Bl -tag -width vfs.root.mountfrom -compact
.It Ev vfs.root.mountfrom
The
.Xr kenv 1
variable that identifies a pool for which the options are written.
.El
.Sh EXAMPLES
Try to boot to a new
.Em boot environment
without changing the
.Cm bootfs
property of a pool:
.Pp
.Dl "zfsbootcfg ""zfs:tank/ROOT/newbe:""
.Pp
To clear the boot options:
.Pp
.Dl "zfsbootcfg """"
.Sh SEE ALSO
.Xr boot.config 5 ,
.Xr bectl 8 ,
.Xr gptzfsboot 8 ,
.Xr loader 8 ,
.Xr zfsboot 8
.Sh HISTORY
.Nm
appeared in
.Fx 11.1 .
.Sh AUTHORS
This manual page was written by
.An Andriy Gapon Aq Mt avg@FreeBSD.org .


sbin/zfsbootcfg/zfsbootcfg.c
/*-
 * Copyright (c) 2016 Andriy Gapon <avg@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <sys/types.h>
#include <errno.h>
#include <limits.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <kenv.h>
#include <unistd.h>

#include <libzfsbootenv.h>

#ifndef ZFS_MAXNAMELEN
#define	ZFS_MAXNAMELEN	256
#endif

int
main(int argc, char * const *argv)
{
	char buf[ZFS_MAXNAMELEN], *name;
	const char *key, *value, *type;
	int rv;
	bool print;

	name = NULL;
	key = NULL;
	type = NULL;
	value = NULL;
	print = false;
	while ((rv = getopt(argc, argv, "k:pt:v:z:")) != -1) {
		switch (rv) {
		case 'k':
			key = optarg;
			break;
		case 'p':
			print = true;
			break;
		case 't':
			type = optarg;
			break;
		case 'v':
			value = optarg;
			break;
		case 'z':
			name = optarg;
			break;
		}
	}

	argc -= optind;
	argv += optind;

	if (argc == 1)
		value = argv[0];

	if (argc > 1) {
		fprintf(stderr, "usage: zfsbootcfg <boot.config(5) options>\n");
		return (1);
	}

	if (name == NULL) {
		rv = kenv(KENV_GET, "vfs.root.mountfrom", buf, sizeof(buf));
		if (rv <= 0) {
			perror("can't get vfs.root.mountfrom");
			return (1);
		}

		if (strncmp(buf, "zfs:", 4) == 0) {
			name = strchr(buf + 4, '/');
			if (name != NULL)
				*name = '\0';
			name = buf + 4;
		} else {
			perror("not a zfs root");
			return (1);
		}
	}

	rv = 0;
	if (key != NULL || value != NULL) {
		if (type == NULL)
			type = "DATA_TYPE_STRING";

		if (key == NULL || strcmp(key, "command") == 0)
			rv = lzbe_set_boot_device(name, value);
		else
			rv = lzbe_set_pair(name, key, type, value);

		if (rv == 0)
			printf("zfs bootenv is successfully written\n");
		else
			printf("error: %d\n", rv);
	} else if (!print) {
		char *ptr;

		if (lzbe_get_boot_device(name, &ptr) == 0) {
			printf("zfs:%s:\n", ptr);
			free(ptr);
		}
	}

	if (print) {
		rv = lzbe_bootenv_print(name, stdout);
	}

	return (rv);
}
